src/envs/behavior_options.py

```diff
@@ -514,48 +510,102 @@ def graspObjectOptionPolicy(_state: State,
         rh_orig_grasp_postion = env.robots[0].parts["right_hand"].get_position(
         )
         rh_orig_grasp_orn = env.robots[0].parts["right_hand"].get_orientation()
 
-        # 1 Move Hand to Grasp Location
+        # 1 Teleport Hand to Grasp offset location
         env.robots[0].parts["right_hand"].set_position_orientation(
             rh_final_grasp_postion,
             p.getQuaternionFromEuler(rh_final_grasp_orn))
 
-        for i in range(25):
-            ret_action = get_delta_low_level_hand_action(
-                env, plan[i - 26][0:3], plan[i - 26][3:6], plan[i - 25][0:3],
-                plan[i - 25][3:6])
-            env.step(ret_action)
+        # 2. Slowly move hand to the surface of the object to be grasped.
+        # We can't simply teleport here since this might be slightly
+        # inside the object!
+        # Use an error-correcting closed-loop!
+        atol_xyz = 1e-4
+        atol_vel = 5e-3
+
+        # Error-correcting closed loop.
+        ec_loop_counter = 0
+        while hand_i < 0 and ec_loop_counter < 60:
+            current_pos = list(
+                env.robots[0].parts["right_hand"].get_position())
+            current_orn = list(
+                p.getEulerFromQuaternion(
+                    env.robots[0].parts["right_hand"].get_orientation()))
+            expected_pos = np.array(plan[hand_i][0:3])
+            # If we're not where we expect in the plan, take some corrective
+            # action
+            if not np.allclose(current_pos, expected_pos, atol=atol_xyz):
+                low_level_action = (get_delta_low_level_hand_action(
+                    env,
+                    np.array(current_pos),
+                    np.array(current_orn),
+                    np.array(plan[hand_i][0:3]),
+                    np.array(plan[hand_i][3:]),
+                ))
+                # if the corrective action is 0, move on
+                if np.allclose(
+                        low_level_action,
+                        np.zeros((17, 1)),
+                        atol=atol_vel,
+                ):
+                    low_level_action = (get_delta_low_level_hand_action(
+                        env,
+                        np.array(current_pos),
+                        np.array(current_orn),
+                        np.array(plan[hand_i + 1][0:3]),
+                        np.array(plan[hand_i + 1][3:]),
+                    ))
+                    hand_i += 1
+                env.step(low_level_action)
+            # Else, move the hand_i pointer to make the expected position
+            # the next position in the plan.
+            else:
+                hand_i += 1
+            ec_loop_counter += 1
 
-        # 2 Simulate Grasp
+        # 3. Close hand and simulate grasp
         a = np.zeros(17, dtype=float)
         a[16] = 1.0
         assisted_grasp_action = np.zeros(28, dtype=float)
         assisted_grasp_action[26] = 1.0
+        if isinstance(obj_to_grasp.body_id, List):
```

Comment:
i'm more used to seeing `isinstance(x, list)` than `isinstance(x, List)`, but I can't think of any reason to prefer the former... hopefully they're equivalent? i googled around for a minute but didn't learn anything \shrug